core-tools.markdown

```diff
@@ -0,0 +1,297 @@
+This will often make the program run much faster and perform more aggressive
+optimisations such as inlining and specialisation.
+For even more optimisations that may take significantly longer to compile,
+`-O2` is also available.
+
+## An interactive environment
+
+GHC provides an interactive environment in a form of a
+Read-Evaluate-Print Loop (REPL) called GHCi.
+To enter the environment run the program `ghci`.
+
+```sh
+➜ ghci
+GHCi, version 9.0.1: https://www.haskell.org/ghc/  :? for help
+ghci> 
+```
+
+It provides an interactive prompt where Haskell expressions can be written and
+evaluated.
+
+For example:
+
+```sh
+ghci> 1 + 1
+2
+ghci> putStrLn "Hello, world!"
+Hello, world!
+```
+
+We can define new names:
+
+```sh
+ghci> double x = x + x
+ghci> double 2
+4
+```
+
+We can write multi-line code by surrounding it with `:{` and `:}`:
+
+```hs
+ghci> :{
+| map f list =
+|     case list of
+|         [] -> []
+|         x : xs -> f x : map f xs
+| :}
+ghci> map (+1) [1, 2, 3]
+[2,3,4]
+
+```
+
+We can import Haskell source files using the `:load` command (`:l` for short):
+
+```sh
+ghci> :load hello.hs
+[1 of 1] Compiling Main             ( hello.hs, interpreted )
+Ok, one module loaded.
+ghci> main
+Hello, Haskell!
+```
+
+As well as import library modules:
+
+```sh
+ghci> import Data.Bits
+ghci> shiftL 32 1
+64
+ghci> clearBit 33 0
+32
+```
+
+We can even ask what the type of an expression is using the `:type` command
+(`:t` for short):
+
+```sh
+λ> :type putStrLn
+putStrLn :: String -> IO ()
+```
+
+To exit `ghci`, use the `:quit` command (or `:q` for short)
+
+```sh
+ghci> :quit
+Leaving GHCi.
+```
+
+A more thorough introduction to GHCi can be found in the
+[GHC user guide](https://downloads.haskell.org/~ghc/9.0.1/docs/html/users_guide/ghci.html).
+
+### Using external packages in ghci
+
+By default, GHCi can only load and use packages that are
+[included with the GHC installation](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/9.0.1-notes.html#included-libraries).
+
+However, users of [cabal-install](https://www.haskell.org/cabal) and
+[stack](http://haskellstack.org) can download and load external packages
+very easily using the following commands:
+
+cabal-install:
+
+```sh
+cabal repl --build-depends async --build-depends say
+```
+
+Stack:
+
+```sh
+stack exec --package async --package say -- ghci
+```
+
+And the modules of the relevant packages will be available for import:
+
+```sh
+GHCi, version 9.0.1: https://www.haskell.org/ghc/  :? for help
+ghci> import Control.Concurrent.Async 
+ghci> import Say
+ghci> concurrently_ (sayString "Hello") (sayString "World")
+Hello
+World
+```
+
+Stack users can also use this feature with `runghc` and `ghc` by replacing
+`ghci` in the command above, and cabal-install users can generate an
+environment file that will make `async` and `say` visible for GHC tools
+in the current directory using this command:
+
+```sh
+cabal install --lib async say --package-env .
```

Comment:
`cabal install --lib` adds stuff to the default environment, which can cause very confusing build errors later when building projects. I'd advise to not mention it at all.

Instead, use `cabal repl --build-depends async`.

Reply:
done. thanks.

Reply:
@hasufell that's not totally correct: the command as was written worked just fine and didn't touch the global environment. That's because the `--package-env .` part. I'd vote for putting it back. But up to you guys.

Reply:
@ulysses4ever Thanks, I overlooked that.

Reply:
I see you both agree that this is actually ok, so I'll revert the change.

Reply:
I also double-checked that preceding text tries to explain what's happening in a nutshell. But maybe Julian has a point here and it would make sense to add a warning (after the command):

> (Warning: the command may work without the `--package-env .` part but in that case it will use the global environment instead of the local one. This is known to lead to complicated problems in some cases and discouraged.)

Of course, the warning itself may sound too confusing for a novice, as well as simply extending the text that we'd like to keep short, so I am not 100% sure.

Reply:
I'd like avoid these kind of warnings in this text.  I also don't think it's relevant because the specific command we teach isn't problematic. If users want to experiment with different commands than what we show they will have to figure it out from other texts.